src/stirling/source_connectors/socket_tracer/uprobe_manager.cc

```diff
@@ -566,6 +587,147 @@ int UProbeManager::DeployOpenSSLUProbes(const absl::flat_hash_set<md::UPID>& pid
+  const std::vector<std::string_view> lib_names = {kGrpcCPythonLibPrefix};
+
+  const system::Config& sysconfig = system::Config::GetInstance();
+
+  // Find path to grpc-c shared object, if it's used (i.e. mapped).
+  PL_ASSIGN_OR_RETURN(const std::vector<std::filesystem::path> container_lib_paths,
+                      FindHostPathForPIDPath(lib_names, pid, proc_parser_.get(), &fp_resolver_,
+                                             HostPathForPIDPathSearchType::kSearchTypeContains));
+
+  std::filesystem::path container_libgrpcc = container_lib_paths[0];
+
+  if (container_libgrpcc.empty()) {
+    // Looks like this process doesn't have dynamic grpc-c library mapped.
+    return 0;
+  }
+
+  // Convert to host path, in case we're running inside a container ourselves.
+  container_libgrpcc = sysconfig.ToHostPath(container_libgrpcc);
+  if (!fs::Exists(container_libgrpcc)) {
+    return error::Internal("grpc-c library not found [path = $0]", container_libgrpcc.string());
+  }
+
+  // Only try probing .so files that we haven't already set probes on.
+  auto result = grpc_c_probed_binaries_.insert(container_libgrpcc);
+  if (!result.second) {
+    return 0;
+  }
+
+  // Calculate MD5 hash of the grpc-c library to know which version it is.
+  // Based on https://stackoverflow.com/questions/1220046/how-to-get-the-md5-hash-of-a-file-in-c
+  PL_ASSIGN_OR_RETURN(const std::string hash_str, MD5onFile(container_libgrpcc.string()));
+  LOG(INFO) << absl::Substitute("Found MD5 hash $0 of library $1", hash_str,
+                                container_libgrpcc.string());
+
+  // Find the version of the library by its MD5 hash.
+  if (kGrpcCMD5HashToVersion.find(hash_str) == kGrpcCMD5HashToVersion.end()) {
+    LOG(INFO) << absl::Substitute("Failed to map hash $0 of gRPC-C library $1 and pid $2", hash_str,
+                                  container_libgrpcc.string(), pid);
+    return error::Unimplemented("Unknown MD5 hash of library.");
+  }
+  const enum grpc_c_version_t version = kGrpcCMD5HashToVersion.at(hash_str);
+  std::unique_ptr<ebpf::BPFHashTable<uint32_t, uint64_t>> grpc_c_versions_map = nullptr;
+  grpc_c_versions_map = std::make_unique<ebpf::BPFHashTable<uint32_t, uint64_t>>(
+      bcc_->GetHashTable<uint32_t, uint64_t>(UProbeManager::kGrpcCVersionsName));
+  LOG(INFO) << absl::Substitute("Updating gRPC-C version of pid $0 to $1", pid, (uint32_t)version);
+  if (!grpc_c_versions_map->update_value(pid, version).ok()) {
+    LOG(WARNING) << absl::Substitute("Failed to update version of pid $0 to $1", pid,
+                                     (uint32_t)version);
+    return error::Internal("Failed to set version of library.");
+  }
+
+  // Attach the needed probes.
+  // This currently works only for non-stripped versions of the shared object.
+  LOG(INFO) << absl::Substitute("Attaching GRPC-C uprobes to $0 for pid $1",
+                                container_libgrpcc.string(), pid);
```

Comment:
Done

Reply:
question for future:
I used LOGs instead of VLOGs because I didn't find where the VLOGs write to, and LOGs simply go to stdout. Where can I see the output of VLOGs?